spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java

```diff
@@ -0,0 +1,180 @@
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.data.datastore.core;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+
+import com.google.cloud.datastore.Datastore;
+import com.google.cloud.datastore.Entity;
+import com.google.cloud.datastore.Entity.Builder;
+import com.google.cloud.datastore.Key;
+import com.google.cloud.datastore.KeyFactory;
+import com.google.cloud.datastore.Query;
+import com.google.common.annotations.VisibleForTesting;
+
+import org.springframework.cloud.gcp.data.datastore.core.convert.DatastoreEntityConverter;
+import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreDataException;
+import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreMappingContext;
+import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastorePersistentEntity;
+import org.springframework.data.mapping.PersistentProperty;
+import org.springframework.util.Assert;
+
+/**
+ * An implementation of {@link DatastoreOperations}.
+ *
+ * @author Chengyuan Zhao
+ *
+ * @since 1.1
+ */
+public class DatastoreTemplate implements DatastoreOperations {
+
+	private final Datastore datastore;
+
+	private final DatastoreEntityConverter datastoreEntityConverter;
+
+	private final DatastoreMappingContext datastoreMappingContext;
+
+	public DatastoreTemplate(Datastore datastore,
+			DatastoreEntityConverter datastoreEntityConverter,
+			DatastoreMappingContext datastoreMappingContext) {
+		Assert.notNull(datastore, "A non-null Datastore service object is required.");
+		Assert.notNull(datastoreEntityConverter,
+				"A non-null DatastoreEntityConverter is required.");
+		Assert.notNull(datastoreMappingContext,
+				"A non-null DatastoreMappingContext is required.");
+		this.datastore = datastore;
+		this.datastoreEntityConverter = datastoreEntityConverter;
+		this.datastoreMappingContext = datastoreMappingContext;
+	}
+
+	@VisibleForTesting
+	Key getKeyFromId(Object id, Class entityClass) {
+		Assert.notNull(id, "Cannot get key for null ID value.");
+		if (id instanceof Key) {
+			return (Key) id;
+		}
+		KeyFactory keyFactory = this.datastore.newKeyFactory();
+		keyFactory.setKind(
+				this.datastoreMappingContext.getPersistentEntity(entityClass).kindName());
+		Key key;
+		if (id instanceof String) {
+			key = keyFactory.newKey((String) id);
+		}
+		else if (id == long.class) {
+			key = keyFactory.newKey((long) id);
+		}
+		else {
+			// We will use configurable custom converters to try to convert other types to
```

Comment:
Why not externalize id conversion into its own interface, so that this becomes more configurable from the start?
You would also not need to use the `@VisibleForTesting` since the logic would be in its own class easily testable.

Reply:
The conversion from other types to the String or Long type indeed would come from outside. I plan to be utilizing Spring Data's custom converters and conversion service classes and making our implementations separately like we did in Spanner and also that is used in Mongodb and older implementations.

However this specific function must use the template's own instance of the Datastore service, and an instance of the mapping context. it might make sense to put it into some sort of util class,  but at this point in the interest of keeping things as simple as needed until we see otherwise i think we should keep this method in this template class because of it's use of 2 things the template holds and also the rather specific Key-related functionality. 

Totally open to moving it out in the future when we have more of these utility style functions though.